SPO-5857 Авторизация и Профиль

ivatar39 opened the pull request:
Имплементирован базовый флоу: 
1. Ввод ФИО.
2. Проверка на то, есть ли юзер с таким именем или нет.
3. Если нет, то создаётся новый. Если есть, то вход по старым данным.
4. Просмотр информации в профиле
5. Выход из аккаунта, смена параметра "isAuthorized" на "false".

ivatar39 committed:
SPO-5857 Auth data layer
- Implemented simple CRUD operations for auth.
- Added repository to perform on lds

ivatar39 committed:
SPO-5857 Auth WMs
- Began work on models and their connections to repositories.

ivatar39 committed:
SPO-5857 Auth and Profile
- All wm's and pages are done
- Complete auth flow has been implemented

saturov commented on .gitignore:
```diff
@@ -44,3 +44,6 @@ app.*.map.json
 /android/app/release
+
+# Temp files and tools
+lib/tools
```
А почему было решено инструменты не коммитить?

ivatar39 replied:
> А почему было решено инструменты не коммитить?

Генератор пришлось переписывать под null-safety и я был очень не уверен местами с null-чеками и правками по линтеру. К тому же, сгенерированный скрин "из коробки" не работал, так как на surf_mwwm используется `ProxyWidgetState`. Пришлось всё равно сгенерированные классы переписывать под чистый mwwm.
Не хотелось всем этим загрязнять репозиторий, скажу честно.

saturov commented on pubspec.yaml:
```diff
@@ -13,25 +13,53 @@ dependencies:
+  uuid: ^3.0.4
+
+  # Routing
+  auto_route: 2.2.0
```
Почему навигацию было решено делать через это? Каких возможностей в стандартном подходе не хватило?

ivatar39 replied:
> Почему навигацию было решено делать через это? Каких возможностей в стандартном подходе не хватило?

Авто-раутинг позволил мне динамически менять скрины и их параметры. Не надо писать много бойлерплейта, всё просто решается кодо-генерацией. В будущем возможно будет проще имплементировать дип-линки и раутинг-обсёрверы для аналитики. 
В стандартном подходе не нравится передача аргументов и довольно бойлерплейтный подход к объявлению всех рутов.

saturov commented on pubspec.yaml:
```diff
@@ -13,25 +13,53 @@ dependencies:
   surf_logger: ^1.0.1
 
+  # Storage
+  hive: ^2.0.4
```
Рассматривал ли ты другие варианты для персистентного хранения данных? Почему остановился на этом?

ivatar39 replied:
> Рассматривал ли ты другие варианты для персистентного хранения данных? Почему остановился на этом?

Раньше я использовал SharedPreferences, но потом понял, что хайв делает то же самое, только лучше. Не нужно мануально прописывать toJson и fromJson методы, авто-генерация делает простой адаптер к классу, что сохраняет время и не принуждает к переписыванию служебных методов каждый раз когда дата-класс меняется. Перформанс ещё, но тут я подробно не мониторил. Авторы просто говорят о хороших [результатах](https://raw.githubusercontent.com/hivedb/hive/master/.github/benchmark_write.png).

saturov commented on lib/ui/splash/splash_wm.dart:
```diff
@@ -0,0 +1,25 @@
+      await _router.replace(AuthPageRoute());
+    }
+    super.onLoad();
+  }
```
В таких функциях как onLoad(), да и в других переопределенных функциях, которые можно назвать функциями жизненного цикла WM, лучше не писать код непосредственно, а лишь осуществлять вызовы других функций. Так код бегло читать станет намного проще.

Вынеси всё это в одну функцию с названием, отражающим суть происходящего.

saturov commented on lib/ui/splash/splash_wm.dart:
```diff
@@ -0,0 +1,25 @@
+import 'package:surf_mwwm/surf_mwwm.dart';
+
+class SplashWidgetModel extends WidgetModel {
+  final AuthRepository _repository;
```
Не выразительное имя. Добавится сюда какой-нибудь другой репозиторий, что делать тогда? Трогать код, который не относится к делу?

ivatar39 replied:
> Не выразительное имя. Добавится сюда какой-нибудь другой репозиторий, что делать тогда? Трогать код, который не относится к делу?

Не совсем понял конкретно какое имя не устраивает. `SplashWidgetModel` или `AuthRepository`?

saturov commented on lib/ui/splash/splash_page.dart:
```diff
@@ -0,0 +1,35 @@
+import 'package:flutter/material.dart';
+import 'package:surf_mwwm/surf_mwwm.dart';
+
+class SplashPage extends CoreMwwmWidget<SplashWidgetModel> {
```
Мы называем экраны не страницами, а всё же экранами обычно. Поэтому не Page, а Screen.

saturov replied:
Мы называем экраны не страницами, а всё же экранами обычно. Поэтому не Page, а Screen.

saturov commented on lib/ui/splash/splash_page.dart:
```diff
@@ -0,0 +1,35 @@
+            const WidgetModelDependencies(),
+            getIt<AuthRepository>(),
+            getIt<AppRouter>(),
+          ),
```
Обычно, мы выносим создание WM в top-level функцию. Тем самым, хотя бы визуально отделяя DI от WM.

saturov commented on lib/ui/router/app_router.dart:
```diff
@@ -0,0 +1,14 @@
+    AdaptiveRoute<dynamic>(page: AuthPage),
+    AdaptiveRoute<dynamic>(page: ProfilePage),
+  ],
+)
```
Расскажи мне про пакет auto_route подробнее. Я им никогда не пользовался. Что он умеет хорошего?

ivatar39 replied:
> Расскажи мне про пакет auto_route подробнее. Я им никогда не пользовался. Что он умеет хорошего?

Описал подробнее [выше](https://github.com/ivatar39/board_manager/pull/1#discussion_r651665457)

saturov commented on lib/ui/profile/profile_wm.dart:
```diff
@@ -0,0 +1,25 @@
+import 'package:surf_mwwm/surf_mwwm.dart';
+
+class ProfileWidgetModel extends WidgetModel {
+  final AuthRepository _repository;
```
Поработать над именованием поля класса. Сейчас оно ни о чем не говорит.

saturov commented on lib/ui/profile/profile_wm.dart:
```diff
@@ -0,0 +1,25 @@
+  @override
+  Future<void> onLoad() async {
+    super.onLoad();
+  }
```
Функции не стоит переопределять, если это не имеет понятного смысла. 
Код "на будущее" складировать в репозитории тоже не нужно. Это верный способ превратить кодовую базу в бардак.

saturov commented on lib/ui/profile/profile_wm.dart:
```diff
@@ -0,0 +1,25 @@
+  Future<void> signOut(User user) async {
+    await _repository.signOut(user);
+    await _router.replace(AuthPageRoute());
+  }
```
Наименование функции не в полной мере соответствует тому, что происходит в её реализации.

saturov commented on lib/ui/profile/profile_page.dart:
```diff
@@ -0,0 +1,56 @@
+import 'package:surf_mwwm/surf_mwwm.dart';
+import 'package:board_manager/ui/router/app_router.gr.dart';
+
+class ProfilePage extends CoreMwwmWidget<ProfileWidgetModel> {
```
Page -> Screen

saturov commented on lib/ui/profile/profile_page.dart:
```diff
@@ -0,0 +1,56 @@
+            const WidgetModelDependencies(),
+            getIt<AuthRepository>(),
+            getIt<AppRouter>(),
+          ),
```
Вынести создание WM в top-level функцию.

saturov commented on lib/ui/profile/profile_page.dart:
```diff
@@ -0,0 +1,56 @@
+
+  @override
+  WidgetState<CoreMwwmWidget<ProfileWidgetModel>, ProfileWidgetModel>
+      createWidgetState() {
```
Странный перенос.

saturov commented on lib/ui/profile/profile_page.dart:
```diff
@@ -0,0 +1,56 @@
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: const Text('Профиль'),
```
На будущее - в любой проект сразу закладывай механизм локализации для строк. Задним числом будет очень сложно интегрировать.

ivatar39 replied:
> На будущее - в любой проект сразу закладывай механизм локализации для строк. Задним числом будет очень сложно интегрировать.

В принципе могу уже сейчас задействовать [библиотеку](https://pub.dev/packages/flappy_translator), которую использую обычно. Много времени не займёт.

saturov commented on lib/ui/auth/auth_wm.dart:
```diff
@@ -0,0 +1,26 @@
+import 'package:surf_mwwm/surf_mwwm.dart';
+
+class AuthWidgetModel extends WidgetModel {
+  final AuthRepository _repository;
```
Поправить именование

saturov commented on lib/ui/auth/auth_wm.dart:
```diff
@@ -0,0 +1,26 @@
+  @override
+  Future<void> onLoad() async {
+    super.onLoad();
+  }
```
Не переопределять функцию без реализации

saturov commented on lib/ui/auth/auth_wm.dart:
```diff
@@ -0,0 +1,26 @@
+    final user = await _repository.registerUser(name);
+    if (user != null) {
+      await _router.replace(ProfilePageRoute(user: user));
+    } else {}
```
Не оставлять пустых логических веток.

saturov commented on lib/ui/auth/auth_wm.dart:
```diff
@@ -0,0 +1,26 @@
+    super.onLoad();
+  }
+
+  Future<void> signIn(String name) async {
```
Привести наименование функции в соответствие с её реальной ответственностью.

saturov commented on lib/ui/app/app.dart:
```diff
@@ -1,18 +1,25 @@
+            borderRadius: BorderRadius.circular(8.0),
+          ),
+        ),
+      ),
```
Тему приложения всегда лучше выносить в другой файл сразу. Она имеет свойство разрастаться.

saturov commented on lib/ui/auth/auth_page.dart:
```diff
@@ -0,0 +1,79 @@
+                autovalidateMode: AutovalidateMode.onUserInteraction,
+                controller: _nameController,
+                validator: (value) {
+                  if (value!.isEmpty || value.length > 80) {
```
80 - магическая константа. Вынести в именованную константу

saturov commented on lib/ui/auth/auth_page.dart:
```diff
@@ -0,0 +1,79 @@
+                    return 'Неверный ввод';
+                  }
+                  return null;
+                },
```
Это не ок, что валидация происходит на UI. Подумай о том, как можно перенести её в WM.

saturov commented on lib/repository/auth/auth_repository.dart:
```diff
@@ -0,0 +1,52 @@
+  Future<User?> registerUser(String name) async {
+    try {
+      final user = User.create(name: name);
+      Logger.d(user.toString());
```
Не оставляем дебажные логи в репозитории. Если они нужны на постоянку, преобразовать их в i, v или другой тип логов.

saturov commented on lib/repository/auth/auth_repository.dart:
```diff
@@ -0,0 +1,52 @@
+      return user;
+    } on AuthException catch (e) {
+      Logger.e(e.toString());
+      return null;
```
Проглатываешь ошибку. Как ты планируешь с учетом этого обработку ошибок реализовать?

saturov commented on lib/repository/auth/auth_repository.dart:
```diff
@@ -0,0 +1,52 @@
+      // Checking if there is any user in data source.
+      // If user exists, we check if they're authorized.
+      final doesAnyUserExist = await _localDataSource.doesAnyUserExist();
+      Logger.d(doesAnyUserExist.toString());
```
Дебажные логи убрать или преобразовать в логи другого типа.

saturov commented on lib/data/auth/data_source/auth_local_data_source.dart:
```diff
@@ -0,0 +1,105 @@
+
+  @override
+  Future<void> saveUser(User user) async {
+    Logger.d('save user: $user');
```
Дебажные логи

saturov commented on lib/data/auth/data_source/auth_local_data_source.dart:
```diff
@@ -0,0 +1,105 @@
+
+  AuthHiveDataSource(
+    this._hive,
+  );
```
Круто, что тут передаешь в конструктор инстанс Hive. Это очень правильный подход.

saturov commented on lib/data/auth/data_source/auth_local_data_source.dart:
```diff
@@ -0,0 +1,105 @@
+  Future<void> saveUser(User user) async {
+    Logger.d('save user: $user');
+    try {
+      final box = _hive.box<User>('users');
```
Эта строка копипастится из функции в функцию. Точно никак нельзя её вынести?

saturov commented on lib/data/auth/data_source/auth_local_data_source.dart:
```diff
@@ -0,0 +1,105 @@
+      Logger.e(e.toString());
+      throw AuthException.memoryException();
+    }
+  }
```
Очень много копипасты. Точно можно причесать класс, чтобы её количество снизить.

ivatar39 committed:
SPO-5857 Auth and Profile rework
- Fixed issues and error handling.
- Styling and linting errors.
- Implemented auto sign-in
